targets/pna_nic/primitives.cpp

```diff
@@ -0,0 +1,362 @@
+#include <bm/bm_sim/phv.h>
+#include <bm/bm_sim/logger.h>
+
+#include <random>
+#include <thread>
+
+namespace bm {
+
+namespace pna {
+
+class modify_field : public ActionPrimitive<Data &, const Data &> {
+  void operator ()(Data &dst, const Data &src) {
+    bm::core::assign()(dst, src);
+  }
+};
+
+REGISTER_PRIMITIVE(modify_field);
+
+class modify_field_rng_uniform
+  : public ActionPrimitive<Data &, const Data &, const Data &> {
+  void operator ()(Data &f, const Data &b, const Data &e) {
+    // TODO(antonin): a little hacky, fix later if there is a need using GMP
+    // random fns
+    using engine = std::default_random_engine;
+    using hash = std::hash<std::thread::id>;
+    static thread_local engine generator(hash()(std::this_thread::get_id()));
+    using distrib64 = std::uniform_int_distribution<uint64_t>;
+    distrib64 distribution(b.get_uint64(), e.get_uint64());
+    f.set(distribution(generator));
+  }
+};
+
+REGISTER_PRIMITIVE(modify_field_rng_uniform);
+
+class add_to_field : public ActionPrimitive<Field &, const Data &> {
+  void operator ()(Field &f, const Data &d) {
+    f.add(f, d);
+  }
+};
+
+REGISTER_PRIMITIVE(add_to_field);
+
+class subtract_from_field : public ActionPrimitive<Field &, const Data &> {
+  void operator ()(Field &f, const Data &d) {
+    f.sub(f, d);
+  }
+};
+
+REGISTER_PRIMITIVE(subtract_from_field);
+
+class add : public ActionPrimitive<Data &, const Data &, const Data &> {
+  void operator ()(Data &f, const Data &d1, const Data &d2) {
+    f.add(d1, d2);
+  }
+};
+
+REGISTER_PRIMITIVE(add);
+
+class subtract : public ActionPrimitive<Data &, const Data &, const Data &> {
+  void operator ()(Data &f, const Data &d1, const Data &d2) {
+    f.sub(d1, d2);
+  }
+};
+
+REGISTER_PRIMITIVE(subtract);
+
+class bit_xor : public ActionPrimitive<Data &, const Data &, const Data &> {
+  void operator ()(Data &f, const Data &d1, const Data &d2) {
+    f.bit_xor(d1, d2);
+  }
+};
+
+REGISTER_PRIMITIVE(bit_xor);
+
+class bit_or : public ActionPrimitive<Data &, const Data &, const Data &> {
+  void operator ()(Data &f, const Data &d1, const Data &d2) {
+    f.bit_or(d1, d2);
+  }
+};
+
+REGISTER_PRIMITIVE(bit_or);
+
+class bit_and : public ActionPrimitive<Data &, const Data &, const Data &> {
+  void operator ()(Data &f, const Data &d1, const Data &d2) {
+    f.bit_and(d1, d2);
+  }
+};
+
+REGISTER_PRIMITIVE(bit_and);
+
+class shift_left :
+  public ActionPrimitive<Data &, const Data &, const Data &> {
+  void operator ()(Data &f, const Data &d1, const Data &d2) {
+    f.shift_left(d1, d2);
+  }
+};
+
+REGISTER_PRIMITIVE(shift_left);
+
+class shift_right :
+  public ActionPrimitive<Data &, const Data &, const Data &> {
+  void operator ()(Data &f, const Data &d1, const Data &d2) {
+    f.shift_right(d1, d2);
+  }
+};
+
+REGISTER_PRIMITIVE(shift_right);
+
+class drop : public ActionPrimitive<> {
+  void operator ()() {
+  }
+};
+
+REGISTER_PRIMITIVE(drop);
+
+class drop_packet : public ActionPrimitive<> {
```

Comment:
remove

Reply:
Will remove for now.

Currently, the PNA implementation only parses the P4 program. When it handles tasks like dropping packets, we must implement `drop_packet` as an extern: https://p4.org/p4-spec/docs/PNA-v0.7.html#sec-extern-function-send_to_port